File: C/Binary Tree - DFS/236. Lowest Common Ancestor of a Binary Tree.c
/* 236. Lowest Common Ancestor of a Binary Tree
Difficulty: Medium
Submission Link: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/submissions/1388953679/ */

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    if (root == NULL || root == p || root == q) {
        return root;
    }

    struct TreeNode *left = lowestCommonAncestor(root->left, p, q);

    struct TreeNode *right = lowestCommonAncestor(root->right, p, q);

    if (left == NULL) {
        return right;
    }
    else if (right == NULL) {
        return left;
    }
    else {
        return root;
    }
}

File: JavaScript/Binary Tree - DFS/236. Lowest Common Ancestor of a Binary Tree.js
/* 236. Lowest Common Ancestor of a Binary Tree
Difficulty: Medium
Submission Link: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/submissions/1388949355/
*/

function lowestCommonAncestor(root, p, q) 
{
    if (!root || root === p || root === q) return root;
    var resL = lowestCommonAncestor(root.left, p, q);
    var resR = lowestCommonAncestor(root.right, p, q);
    return (resL && resR) ? root : (resL || resR);
}